### PR TITLE
Probe flexibility of loss, callbacks and optimize steps

### LINKED ISSUE #18: Refactor optimize step
### Describe the feature

Issue coming from qadence. 
https://github.com/pasqal-io/qadence/issues/732
https://github.com/pasqal-io/qadence/issues/631

### It should be implemented because

_No response_

### Additional context

_No response_

### Would you like to work on this issue?

None

### PR DESCRIPTION
**Pull request by debrevitatevitae**
Closes #18 

## Disclaimer
This PR is not really a refactoring of optimize step, but an exploration of how to satisfy the requirements in #18 with perceptrain.

Rather than fully arbitrary optimize steps, hard to generalize, this PR shows (with examples) how to divide responsibility between losses and callbacks to obtain custom behaviour.

### CODE REVIEW

**Review by DanieleCucurachi — APPROVED** [review]
Great work @debrevitatevitae !

The refactoring looks good, different functionalities are now properly delegated to the appropriate objects, and `optimize_step()` is no longer overloaded.

I also really like the new approach to handling losses and loss balancing techniques. Implementing the balancing logic within the loss class itself, rather than inside `optimize_step()`, inherently avoids the issue we encountered before where the model optimization was done using the original loss, but the returned loss value was the rebalanced loss, leading to misleading results. This is a much cleaner and more robust design.

I also tested R3 sampling on a few simple problem instances, manually verifying the behavior was the expected one, checking all the numbers one by one. It appears to be working exactly as intended.

File: perceptrain/data.py [hunk 1]
```diff
@@ -89,6 +90,92 @@ def __iter__(self) -> Iterator:
                 yield tuple(t[idx] for t in self.tensors)
 
 
+class R3Dataset(Dataset):
+    def __init__(
+        self, proba_dist: Callable[[int], Tensor], n_samples: int, release_threshold: float = 0.1
+    ) -> None:
+        """Dataset for R3 sampling (introduced in https://arxiv.org/abs/2207.02338#).
+
```
[the remark this answers is not in the document]
Comment (debrevitatevitae) [reply perceptrain/data.py:98]: Done indeed.

File: .gitignore [hunk 2]
```diff
@@ -79,3 +79,9 @@ events.out.tfevents.*
 # mlflow
 mlruns/
 mlartifacts/
+
+# pyright
+pyrightconfig.json
+
+# sketches and notes
```
Comment (mlahariya) [inline .gitignore:86]: why is this added here?

File: perceptrain/callbacks/callback.py [hunk 3]
```diff
@@ -803,3 +855,53 @@ def run_callback(self, trainer: Any, config: TrainConfig, writer: BaseWriter) ->
...
+        called_every: int = 1,
+    ):
+        """Callback for R3 sampling (https://arxiv.org/abs/2207.02338#).
+
+        Args:
+            initial_dataset (R3Dataset): The dataset updating according to the R3 scheme.
+            fitness_function (Callable[[Tensor, nn.Module], Tensor]): The function to
+                compute fitness scores for samples. Based on the fitness scores, the
+                samples are retained or released.
+            verbose (bool, optional): Whether to print the callback's summary.
```
Comment (mlahariya) [inline perceptrain/callbacks/callback.py:875]: It will be good to add an example of the fitness function in the docstrings

File: docs/tutorials/callbacks/trainer_callbacks.md [hunk 4]
```diff
@@ -20,20 +20,20 @@ config = TrainConfig(
 The `Trainer` class in `perceptrain` provides built-in support for executing callbacks at various stages in the training process, managed through a callback manager. By default, several callbacks are added to specific hooks to automate common tasks, such as check-pointing, metric logging, and model tracking.
```
Comment (mlahariya) [inline docs/tutorials/callbacks/trainer_callbacks.md:20]: It will be a good idea to write a few lines to provide difference in LivePlot and writeplot callbacks. 

Also in the default callbacks, please add the sections for all the new callbacks we have added. (R3, liveplot etc)

File: docs/tutorials/callbacks/default.md [hunk 5]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (mlahariya) [inline docs/tutorials/callbacks/default.md:1]: In the documentation of data sets (where we provide info about data/dictdataloader) - lets also add documentation about R3 dataset?

File: perceptrain/models.py [hunk 6]
```diff
@@ -46,3 +48,74 @@ def forward(self, x: Tensor) -> Tensor:
         passes x through a classical linear layer.
         """
         return x
+
+
+class FFNN(nn.Module):
+    def __init__(self, layers: Sequence[int], activation_function: nn.Module = nn.GELU()) -> None:
+        """
+        Standard feedforward neural network.
```
Comment (mlahariya) [inline perceptrain/models.py:56]: Thats quite an odd way to define a FFNN. we typically should use num_inputs/num_outputs/num_hidden to define the layers internally.

File: perceptrain/models.py [hunk 7]
```diff
@@ -46,3 +48,74 @@ def forward(self, x: Tensor) -> Tensor:
...
+        self,
+        nn: nn.Module,
+        equations: dict[str, Callable[[Tensor, nn.Module], Tensor]],
+    ) -> None:
+        """Physics-informed neural network.
+
+        Args:
+            nn (nn.Module): Neural network module.
+            equations (dict[str, Callable[[Tensor, nn.Module], Tensor]]): Dictionary of equations.
+                These are assumed in the form LHS(x) = 0, so each term of `equations` should
```
Comment (mlahariya) [inline perceptrain/models.py:102]: Provide a simple example for the equations here.

File: perceptrain/callbacks/callback.py (hunk 3 above)
Comment (debrevitatevitae) [reply perceptrain/callbacks/callback.py:875]: Done (8f3c4da)

File: docs/tutorials/callbacks/default.md (hunk 5 above)
Comment (debrevitatevitae) [reply docs/tutorials/callbacks/default.md:1]: Done (1c0a941)

File: perceptrain/models.py (hunk 6 above)
Comment (debrevitatevitae) [reply perceptrain/models.py:56]: It's just more synthetic.

If you passed `num_inputs`, `num_outputs` and `num_hidden` separately, then you'd also have to pass a list of `hidden_layers` with the sizes of the hidden layers and make sure that `len(hidden_layers) == num_hidden`.

File: perceptrain/models.py (hunk 7 above)
Comment (debrevitatevitae) [reply perceptrain/models.py:102]: Done (6637052)
